From 3c01ad4d6a4b591f1d75406747e165b1e4492adc Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 12 Apr 2022 13:41:13 +0200
Subject: [PATCH 194/841] refactor(st-bsec): add bsec_get_secure_state

Add a function to get the BSEC secure state, associated to
the global state used in result of BSEC SVC.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git c90ad91a63b267fac7df8e1a5f255fe1185bd3ca refactor(st-bsec): add bsec_get_secure_state]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I962537f919601cc415d1a9f2f7f66e618dc0a30e
---
 drivers/st/bsec/bsec2.c   | 22 ++++++++++++++++++++++
 include/drivers/st/bsec.h | 24 +++++++++++++++++++++++-
 2 files changed, 45 insertions(+), 1 deletion(-)

diff --git a/drivers/st/bsec/bsec2.c b/drivers/st/bsec/bsec2.c
index e57e13caf..c88083284 100644
--- a/drivers/st/bsec/bsec2.c
+++ b/drivers/st/bsec/bsec2.c
@@ -859,3 +859,25 @@ uint32_t bsec_check_nsec_access_rights(uint32_t otp)
 	return BSEC_OK;
 }
 
+uint32_t bsec_get_secure_state(void)
+{
+	uint32_t status = bsec_get_status();
+	uint32_t result = BSEC_STATE_INVALID;
+
+	if ((status & BSEC_MODE_INVALID_MASK) != 0U) {
+		result = BSEC_STATE_INVALID;
+	} else {
+		if ((status & BSEC_MODE_SECURE_MASK) != 0U) {
+			if (stm32mp_is_closed_device()) {
+				result = BSEC_STATE_SEC_CLOSED;
+			} else {
+				result = BSEC_STATE_SEC_OPEN;
+			}
+		} else {
+			/* OTP modes OPEN1 and OPEN2 are not supported */
+			result = BSEC_STATE_INVALID;
+		}
+	}
+
+	return result;
+}
diff --git a/include/drivers/st/bsec.h b/include/drivers/st/bsec.h
index cff961839..aae36c093 100644
--- a/include/drivers/st/bsec.h
+++ b/include/drivers/st/bsec.h
@@ -33,6 +33,22 @@
 #define BSEC_NOT_SUPPORTED		0xFFFFFFF7U
 #define BSEC_WRITE_LOCKED		0xFFFFFFF6U
 
+/*
+ * get BSEC global state: result for bsec_get_secure_state()
+ * @state: global state
+ *           [1:0] BSEC state
+ *             00b: Sec Open
+ *             01b: Sec Closed
+ *             11b: Invalid
+ *           [8]: Hardware Key set = 1b
+ */
+#define BSEC_STATE_SEC_OPEN		U(0x0)
+#define BSEC_STATE_SEC_CLOSED		U(0x1)
+#define BSEC_STATE_INVALID		U(0x3)
+#define BSEC_STATE_MASK			GENMASK_32(1, 0)
+
+#define BSEC_HARDWARE_KEY		BIT(8)
+
 /*
  * OTP MODE
  */
@@ -80,8 +96,14 @@ uint32_t bsec_set_sp_lock(uint32_t otp);
 uint32_t bsec_read_sp_lock(uint32_t otp, bool *value);
 uint32_t bsec_read_permanent_lock(uint32_t otp, bool *value);
 uint32_t bsec_otp_lock(uint32_t service);
-bool bsec_mode_is_closed_device(void);
 uint32_t bsec_shadow_read_otp(uint32_t *otp_value, uint32_t word);
+
+uint32_t bsec_get_secure_state(void);
+static inline bool bsec_mode_is_closed_device(void)
+{
+	return (bsec_get_secure_state() & BSEC_STATE_MASK) == BSEC_STATE_SEC_CLOSED;
+}
+
 uint32_t bsec_check_nsec_access_rights(uint32_t otp);
 
 #endif /* BSEC_H */
-- 
2.47.2

